src/main/scala/fi/oph/kouta/util/json.scala

```diff
@@ -6,139 +6,157 @@ import java.util.UUID
 
 import fi.oph.kouta.domain._
 import fi.oph.kouta.domain.oid._
+import fi.oph.kouta.domain.valintaperuste._
 import org.json4s.JsonAST.{JObject, JString}
 import org.json4s.jackson.Serialization.write
-import org.json4s.{CustomKeySerializer, CustomSerializer, DefaultFormats, Formats, Extraction}
+import org.json4s.{CustomKeySerializer, CustomSerializer, DefaultFormats, Extraction, Formats}
 
 import scala.util.Try
 
 trait KoutaJsonFormats extends DefaultKoutaJsonFormats {
 
   implicit def jsonFormats: Formats = koutaJsonFormats
 
-  def toJson(data:AnyRef) = write(data)
+  def toJson(data: AnyRef): String = write(data)
 }
 
 sealed trait DefaultKoutaJsonFormats {
 
-  val ISO_LOCAL_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm")
+  val ISO_LOCAL_DATE_TIME_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm")
 
-  def genericKoutaFormats: Formats = DefaultFormats +
-    new CustomSerializer[Julkaisutila](formats => ( {
-      case JString(s) => Julkaisutila.withName(s)
-    }, {
-      case j: Julkaisutila => JString(j.toString)
-    })) +
-    new CustomSerializer[Koulutustyyppi](formats => ( {
-      case JString(s) => Koulutustyyppi.withName(s)
-    }, {
-      case j: Koulutustyyppi => JString(j.toString)
-    })) +
-    new CustomKeySerializer[Kieli](formats => ( {
-      case s: String => Kieli.withName(s)
-    }, {
-      case k: Kieli => k.toString
-    })) +
-    new CustomSerializer[Hakulomaketyyppi](formats => ({
-      case JString(s) => Hakulomaketyyppi.withName(s)
+  def koutaJsonFormats: Formats = genericKoutaFormats ++ Seq(
+    koulutusMetadataSerializer,
+    toteutusMetadataSerializer,
+    valintatapaSisaltoSerializer,
+    valintaperusteMetadataSerializer)
+
+  private def genericKoutaFormats: Formats = DefaultFormats
+    .addKeySerializers(Seq(kieliKeySerializer)) ++
+    Seq(
+      julkaisuTilaSerializer,
+      koulutusTyyppiSerializer,
+      hakulomaketyyppiSerializer,
+      localDateTimeSerializer,
+      kieliSerializer,
+      uuidSerializer,
+      liitteenToimitustapaSerializer,
+      hakuOidSerializer,
+      hakuKohdeOidSerializer,
+      koulutusOidSerializer,
+      toteutusOidSerializer,
+      organisaatioOidSerializer,
+      userOidSerializer,
+      oidSerializer)
+
+  private def kieliKeySerializer = new CustomKeySerializer[Kieli](_ => ( {
+    case s: String => Kieli.withName(s)
+  }, {
+    case k: Kieli => k.toString
+  }))
+
+  private def stringSerializer[A: Manifest](construct: String => A) = new CustomSerializer[A](_ => ({
+      case JString(s) => construct(s)
     }, {
-      case j: Hakulomaketyyppi => JString(j.toString)
-    })) +
-    new CustomSerializer[LocalDateTime](formats => ({
+      case a: A => JString(a.toString)
+  }))
+
+  private def julkaisuTilaSerializer:         CustomSerializer[Julkaisutila]         = stringSerializer(Julkaisutila.withName)
```

Comment:
Näitä muuttujia ei käytetä kuin yhdessä paikassa, joten ehkä niitä ei tarvitse ollenkaan. Voisi laittaa suoraan: 

private def genericKoutaFormats: Formats = DefaultFormats
    .addKeySerializers(Seq(kieliKeySerializer)) ++
    Seq(
      stringSerializer(Julkaisutila.withName),
      stringSerializer(Koulutustyyppi.withName),
      stringSerializer(Hakulomaketyyppi.withName) ....

Reply:
Mietin tuota, mutta en tehnyt niin, kun kaikki eivät ole noita yhden rivin tapauksia. Noita pidempia serialisoijia on tosin vaan yksi, joten ehkä jäten sen ainoana metodiksi, ja siirrän muiden määrittelyt suoraan tuohon listaan.